### PR TITLE
(api) Add option to enforce use of scopes (`all/read`, `all/write`, etc.) [after #448]

### PR DESCRIPTION
**Pull request by Leobouloc**
# Merge after unification & test refactoring [#448 ]

## Purpose

The current state of Ralph allows to restrict users by authority, but does not allow 1. An admin user 2. Finer access control (read, write). This PR aims to solve this issue by implementing `scopes` (a field already present in user accounts) which can allow restricted access.

## Proposal

The scopes proposed are a slight variation on the scopes defined by the [xAPI standard](https://github.com/adlnet/xAPI-Spec/blob/master/xAPI-Communication.md#details-15):

```
    "statements/write",
    "statements/read/mine",
    "statements/read",
    "state/write",
    "state/read",
    "define",
    "profile/write",
    "profile/read",
    "all/read",
    "all",
```

NB: This PR also proposes some cleaning (renaming tests and factorizing code in tests).

TODO:
- [x] restrict endpoint access by scopes
- [x] deal with  `statements/read/mine` scope (ENFORCE_AUTHORITY must be used) (specific test !)
- [x] test restrictions for all endpoints (GET, POST, PUT) AND for both auth methods (OIDC, and basic auth)
- [x] (side quest) rename tests removing extra "statement" (i.e. `test_api_statements_get_statements` -> `test_api_statements_get`)
- [x] (side quest) create a `mock_statement` helper to simplify post and put tests

### CODE REVIEW

File: tests/helpers.py [hunk 1]
```diff
@@ -121,3 +124,61 @@ def create_mock_agent(
         return agent
 
     raise ValueError("No valid ifi was provided to create_mock_agent")
+
+
+def mock_statement(
```
Comment (wilbrdt) [inline tests/helpers.py:129]: I'm fine with this! We could ask @quitterie-lcs @jmaupetit

**Review by wilbrdt — COMMENTED** [review]
Alt text

File: tests/api/test_statements_get.py [hunk 2]
```diff
@@ -730,6 +735,168 @@ def test_api_statements_get_statements_invalid_query_parameters(
...
+    )
+
+    assert response.status_code == 200
+
+    if read_all_access:
+        assert response.json() == {"statements": [statements[1], statements[0]]}
+    else:
+        assert response.json() == {"statements": [statements[0]]}
+
+    app.dependency_overrides.pop(get_authenticated_user, None)
```
[the remark this answers is not in the document]
Comment (Leobouloc) [reply tests/api/test_statements_get.py:902]: Thank you ahah

File: src/ralph/api/routers/statements.py [hunk 3]
```diff
@@ -328,12 +327,14 @@ async def get(
...
         if settings.LRS_RESTRICT_BY_SCOPES:
-            raise NotImplementedError("Scopes are not yet implemented in Ralph.")
-
-        # Otherwise, enforce mine for all users
-        mine = True
+            if not current_user.scopes.is_authorized("statements/read"):
+                mine = True
+        else:
+            # If not using scopes, enforce "mine" for all users
+            mine = True
```
[the remark this answers is not in the document]
Comment (Leobouloc) [reply src/ralph/api/routers/statements.py:337]: yes good point.

also changing default value of `mine`: 

`mine=settings.LRS_RESTRICT_BY_AUTHORITY`

File: tests/helpers.py (hunk 1 above)
Comment (quitterie-lcs) [reply tests/helpers.py:129]: Sounds good for me too

**Comment by Leobouloc** [comment]
Replaced by #463

**Pull request closed without merging**